WordPress/src/main/java/org/wordpress/android/util/PerAppLocaleManager.kt

```diff
@@ -0,0 +1,132 @@
+        return (locales.isEmpty || locales == LocaleListCompat.getEmptyLocaleList())
+    }
+
+    /**
+     * We want to make sure the language pref for the in-app locale (old implementation) is set
+     * to the same locale as the AndroidX per-app locale. This way LocaleManager.getLanguage -
+     * which is used throughout the app - returns the correct language code. We can remove
+     * this once the per-app language pref is no longer experimental.
+     */
+    private fun checkOldLanguagePrefKey() {
+        val prefKey = LocaleManager.getLocalePrefKeyString()
+        val inAppLanguage = appPrefsWrapper.prefs().getString(prefKey, "")
+        val perAppLanguage = getCurrentLocale().language
+        if (perAppLanguage.isNotEmpty() && inAppLanguage.equals(perAppLanguage).not()) {
+            appPrefsWrapper.prefs().edit().putString(prefKey, perAppLanguage).apply()
+            appLogWrapper.d(
+                AppLog.T.SETTINGS,
+                "PerAppLocaleManager: changed inAppLanguage from $inAppLanguage to $perAppLanguage"
+            )
+        }
+    }
+
+    /*
+     * Useful during testing to clear the system stored app locale
+     */
+    @Suppress("unused")
+    fun resetApplicationLocale() {
+        AppCompatDelegate.setApplicationLocales(LocaleListCompat.getEmptyLocaleList())
+    }
+
+    fun setCurrentLocaleByLanguageCode(languageCode: String) {
+        // We shouldn't have to replace "_" with "-" but this is in order to work with our existing language picker
+        // on pre-Android 13 devices
+        val appLocale = LocaleListCompat.forLanguageTags(languageCode.replace("_", "-"))
+        AppCompatDelegate.setApplicationLocales(appLocale)
+        checkOldLanguagePrefKey()
+    }
+
+    /**
+     * Previously the app locale was stored in SharedPreferences, so here we migrate to AndroidX per-app language prefs
+     */
+    fun performMigrationIfNecessary() {
+        if (isPerAppLanguagePrefsEnabled()) {
+            if (isApplicationLocaleEmpty()) {
+                val prefKey = LocaleManager.getLocalePrefKeyString()
+                val previousLanguage = appPrefsWrapper.prefs().getString(prefKey, "")
+                if (previousLanguage?.isNotEmpty() == true) {
+                    appLogWrapper.d(
+                        AppLog.T.SETTINGS,
+                        "PerAppLocaleManager: performing migration to AndroidX per-app language prefs"
+                    )
+                    setCurrentLocaleByLanguageCode(previousLanguage)
+                } else {
+                    appLogWrapper.d(
+                        AppLog.T.SETTINGS,
+                        "PerAppLocaleManager: setting default locale"
+                    )
+                    setCurrentLocaleByLanguageCode(Locale.getDefault().language)
+                }
+            } else {
+                checkOldLanguagePrefKey()
+            }
+        }
+    }
+
+    fun isPerAppLanguagePrefsEnabled(): Boolean {
```

Comment:
The per-app locale **can** be enabled on pre-13 devices. The version check in `handleAppLocalePickerClick` is to determine whether we should show the system app language settings or our own custom language picker. The former is only supported on 13+ devices.

Reply:
Ah, apologies for this comment Nick, I think you are right, and after me reading this comment of yours:

> Do you mean we should hide the per-app language preference for Android 12 or lower? If so, we want it to be available because thanks to AndroidX, pre-13 devices support the per-app preference, they just don't support the language selection in the system app settings. The goal here is to eventually drop the in-app setting entirely (apart from our custom language picker for pre-13 devices).

Plus remembering reading [this](https://developer.android.com/guide/topics/resources/app-languages):

> For backward compatibility with previous Android versions, equivalent APIs are also available in AndroidX. However, the backward compatible APIs work with the AppCompatActivity context, not the application context, for Android 12 (API level 32) and earlier. Access the backward compatible APIs with [Appcompat 1.6.0](https://developer.android.com/jetpack/androidx/releases/appcompat#version_160_3) or higher.

It is now more clear to me why we would want the `isPerAppLanguagePrefsEnabled()` check happening for Android or lower as well. Then, only supporting the language selection in the system app settings for Android 13 and after, we only want this `Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU` check there, nowhere else.

Once, apologies for the noise... 🫣